Added tests for unsigned blending equations

polymetal0 opened the pull request:
Also passed src and dst factors as variables to method Body.

abaire commented on src/tests/blend_tests.cpp:
```diff
@@ -30,11 +35,33 @@ static constexpr uint32_t kTexturePitch = kTextureSize * 4;
 
 BlendTests::BlendTests(TestHost &host, std::string output_dir, const Config &config)
     : TestSuite(host, std::move(output_dir), "Blend tests", config) {
+  tests_[kAddName] = [this]() {
```
Since you're extending the test function to allow non-ONE factors, let's update the test names to prepare for a larger test matrix in the future. See https://github.com/abaire/nxdk_pgraph_tests/blob/e2b301a68c020813d094abccb26afe4236926b7b/src/tests/fog_tests.cpp#L40 (or anything wth `MakeTestName`) as an example.

Keep in mind that filenames on FATX have a max length of 42 (4 of which will be ".png"; might want to use shorter names for the function like in https://github.com/abaire/nxdk_pgraph_tests/blob/e2b301a68c020813d094abccb26afe4236926b7b/src/tests/texture_signed_component_tests.cpp#L25 and either a similar set of shortnames for blend factors or their hex values.



The test also only applies the blend function to the alpha channel under the assumption that the overall behavior would be the same for all channels. I don't recall if that was ever verified, so it'd be good to get someone with hardware to test and then update the comment on line 233 below if it's clear that the behavior is the same.

polymetal0 replied:
Would the latest changes be a correct approach?

polymetal0 commented:
Right now, only `ONE` factors are applied to both src and dst. But I have some doubts on how they would work together with each other.

- Can we assume that sfactor = dfactor? and if not, do all sfactors combine with all dfactors and vice versa?
- Would src and dst need separate maps?

My intention was to add two nested loops for all possible src and dst factors, if that is correct.

abaire commented on src/tests/blend_tests.cpp:
```diff
@@ -22,47 +22,58 @@ static constexpr uint32_t kDefaultDMAColorChannel = 9;
+    name += "_";
+    name += kBlendFactors[1].name;
+    name += "_";
+    name += test_eqn.name;
```
I think it'd be more readable to have the equation first or in between the src and dst factors.

abaire commented on src/tests/blend_tests.cpp:
```diff
@@ -22,47 +22,58 @@ static constexpr uint32_t kDefaultDMAColorChannel = 9;
-        NV097_SET_BLEND_FUNC_SFACTOR_V_ONE, NV097_SET_BLEND_FUNC_DFACTOR_V_ONE); });
-  };
+  for (const auto &test_eqn : kBlendEqns) {
+    std::string name = "";
```
Nit, let's just assign the first component of the string directly instead of constructing an empty one and immediately concatting.

abaire commented on src/tests/blend_tests.cpp:
```diff
@@ -22,47 +22,58 @@ static constexpr uint32_t kDefaultDMAColorChannel = 9;
+    {"dstRGB", NV097_SET_BLEND_FUNC_SFACTOR_V_DST_COLOR},
+    {"1-dstRGB", NV097_SET_BLEND_FUNC_SFACTOR_V_ONE_MINUS_DST_COLOR},
+    {"srcAsat", NV097_SET_BLEND_FUNC_SFACTOR_V_SRC_ALPHA_SATURATE},
+    {"ctRGB", NV097_SET_BLEND_FUNC_SFACTOR_V_CONSTANT_COLOR},
```
Nit: Let's change all the "ct"'s to just "c"'s.

abaire commented on src/tests/blend_tests.cpp:
```diff
@@ -22,20 +22,58 @@ static constexpr uint32_t kDefaultDMAColorChannel = 9;
+    name += "_";
+    name += test_eqn.name;
+    tests_[name] = [this, name, &test_eqn, factor]() {
+      Test(test_eqn.name, [this, name, &test_eqn, factor]() { 
```
Even though they're the same now, let's change everything to pass source and destination factors separately.

polymetal0 replied:
this is wrong; should be `Test(name,...`

abaire commented:
> Right now, only `ONE` factors are applied to both src and dst. But I have some doubts on how they would work together with each other.
> 
> * Can we assume that sfactor = dfactor? and if not, do all sfactors combine with all dfactors and vice versa?
> * Would src and dst need separate maps?
> 
> My intention was to add two nested loops for all possible src and dst factors, if that is correct.

No, if we want the test to be comprehensive we shouldn't assume source and destination factors are equal. There are certainly combinations that don't make sense (e.g., both being zero), but those could be filtered out to reduce the number of permutations. In practice we might want to just cherry-pick a few interesting combinations and assume that any major regression (e.g., failure to apply one of the factors at all) would break so many things that it'd be caught without the help of these tests.


As an aside, please make sure you're testing everything locally before asking for reviews. I don't mind terribly, but the xemu reviewer pool is tiny and you'll get a much better reception if you can build a reputation for very high quality, well vetted commits. I'd also suggest refraining from merging from main during reviews, unless your reviewer asks you to. It adds noise and makes it harder for folks who have very limited review time to focus on the parts that matter.

abaire commented on src/tests/blend_tests.cpp:
```diff
@@ -22,20 +22,81 @@ static constexpr uint32_t kDefaultDMAColorChannel = 9;
+    {"1-cA", NV097_SET_BLEND_FUNC_SFACTOR_V_ONE_MINUS_CONSTANT_ALPHA},
+};
+
+static constexpr struct BlendFunc kBlendFactorsDST[] = {
```
There's no reason to duplicate the array. Either these should all be changed to use the `DFACTOR` constants or we can just make a note that SFACTOR and DFACTOR values are all identical and use one array. I'd suggest the latter (you can add a comment above the single array and go back to calling it `kBlendFactors`), since memory is at a premium on the target system.

polymetal0 replied:
Sorry, I didn't notice I didn't swap the Ss for Ds. I used one array as you suggested, but are they actually identical?

abaire replied:
Yes, but you should definitely double check by looking them up where they're defined. In case you're not using an IDE where you can jump to the definitions (you definitely should), they're in the nxdk pbkit code.

polymetal0 replied:
Just to be sure, you mean this? 

![image](https://github.com/user-attachments/assets/728443ec-5297-4960-9933-c29cd3bce848)

abaire replied:
Yes, exactly. As long as all of those indented values are identical between the two lists, it should be fine to use a single array.

abaire commented on src/tests/blend_tests.cpp:
```diff
@@ -22,20 +22,81 @@ static constexpr uint32_t kDefaultDMAColorChannel = 9;
+  }
 }
 
 void BlendTests::Initialize() {
```
Now that the tests are more complicated I think it'd be good to also clarify what is being rendered in the message printed to the screen. On line 173, can you put "Alpha: " before the "G: 0x7F" so it's clear that those values are talking about the alpha value for the various colored boxes?

abaire commented:
Generally looks good, thanks!

To wrap things up:

1. Can you make sure that clang-format was run over the changes? Some of the line breaks seem odd.
2. Once you've verified that the source and destination factor constants are the same, can you squash all the changes and rebase to HEAD?

polymetal0 commented:
> Generally looks good, thanks!
> 
> To wrap things up:
> 
> 1. Can you make sure that clang-format was run over the changes? Some of the line breaks seem odd.
> 2. Once you've verified that the source and destination factor constants are the same, can you squash all the changes and rebase to HEAD?

1. I had autocrlf set to true, is that what you mean? if I set it to false and make another commit will it be enough?
2. I think I have, left a snapshot above

abaire commented on src/tests/blend_tests.cpp:
```diff
@@ -22,20 +22,64 @@ static constexpr uint32_t kDefaultDMAColorChannel = 9;
+        if (sfactor != NV097_SET_BLEND_FUNC_SFACTOR_V_ZERO ||
+            dfactor != NV097_SET_BLEND_FUNC_SFACTOR_V_ZERO) {
+          tests_[name] = [this, name, &test_eqn, sfactor, dfactor]() {
+            Test(name, [this, name, &test_eqn, sfactor, dfactor]() { Body(test_eqn.value, sfactor, dfactor); });
```
nit: Looks like `name` is not used by the inner capture and can be removed (sorry for not noticing before)

polymetal0 replied:
Indeed, removed 👍

abaire commented:
> > Generally looks good, thanks!
> > To wrap things up:
> > 
> > 1. Can you make sure that clang-format was run over the changes? Some of the line breaks seem odd.
> > 2. Once you've verified that the source and destination factor constants are the same, can you squash all the changes and rebase to HEAD?
> 
> 1. I had autocrlf set to true, is that what you mean? if I set it to false and make another commit will it be enough?

No, there are a few places where lines are wrapped too early, the line endings are correct. It's possibly related to the specific version of clang-format you're using (I'm using 19). I added a pre-commit yesterday that fails, if you check the details you can see the remaining issues.

> 2. I think I have, left a snapshot above

Thanks! I agree that the constants are all the same so we should be good with one array.

polymetal0 commented:
> > > Generally looks good, thanks!
> > > To wrap things up:
> > > 
> > > 1. Can you make sure that clang-format was run over the changes? Some of the line breaks seem odd.
> > > 2. Once you've verified that the source and destination factor constants are the same, can you squash all the changes and rebase to HEAD?
> > 
> > 
> > 
> > 1. I had autocrlf set to true, is that what you mean? if I set it to false and make another commit will it be enough?
> 
> No, there are a few places where lines are wrapped too early, the line endings are correct. It's possibly related to the specific version of clang-format you're using (I'm using 19). I added a pre-commit yesterday that fails, if you check the details you can see the remaining issues.
> 
> > 2. I think I have, left a snapshot above
> 
> Thanks! I agree that the constants are all the same so we should be good with one array.

I addressed the issues that showed up in details. I'll tidy things up as soon as all pre-commits are passed.

Btw, I tried squashing and rebasing to main but I couldn't separate my commits from the updates and merges (again my bad, sorry for that). I don't want to mess things up, should I do it manually and send another PR?

abaire commented:
> > > > Generally looks good, thanks!
> > > > To wrap things up:
> > > > 
> > > > 1. Can you make sure that clang-format was run over the changes? Some of the line breaks seem odd.
> > > > 2. Once you've verified that the source and destination factor constants are the same, can you squash all the changes and rebase to HEAD?
> > > 
> > > 
> > > 
> > > 1. I had autocrlf set to true, is that what you mean? if I set it to false and make another commit will it be enough?
> > 
> > 
> > No, there are a few places where lines are wrapped too early, the line endings are correct. It's possibly related to the specific version of clang-format you're using (I'm using 19). I added a pre-commit yesterday that fails, if you check the details you can see the remaining issues.
> > > 2. I think I have, left a snapshot above
> > 
> > 
> > Thanks! I agree that the constants are all the same so we should be good with one array.
> 
> I addressed the issues that showed up in details. I'll tidy things up as soon as all pre-commits are passed.

Looks like the pre-commits found no new errors, so all good there.

> Btw, I tried squashing and rebasing to main but I couldn't separate my commits from the updates and merges (again my bad, sorry for that). I don't want to mess things up, should I do it manually and send another PR?

No, unless a reviewer explicitly asks for a new PR you should almost never create a new one. It's sometimes important to keep the history of the review around to understand why certain decisions were made.

In this case, if you sync your main branch to upstream main, you should be able to `git rebase -i main` and squash the commits into one (I just pulled your branch to verify that this works). Then you can force push to update this PR (since I'm done reviewing and have specifically asked for a squash). 
I think if I do a squash merge when accepting the PR it would also do the right thing, but it's good practice to squash it yourself, since you'll be asked to do this for xemu contributions and it gives you an opportunity to update the commit message if needed.

polymetal0 commented:
I tried rebasing from browser, but it's wrong...I'll better wait for your feedback.

abaire commented:
> > In this case, if you sync your main branch to upstream main, you should be able to `git rebase -i main` and squash the commits into one (I just pulled your branch to verify that this works). Then you can force push to update this PR (since I'm done reviewing and have specifically asked for a squash). I think if I do a squash merge when accepting the PR it would also do the right thing, but it's good practice to squash it yourself, since you'll be asked to do this for xemu contributions and it gives you an opportunity to update the commit message if needed.
> 
> Sorry for taking my time, I squashed and merged the changes into my main and everything appears to be OK, but I'm not being able to force push my main into this PR (I'm using GitHub Desktop so sorry if I'm asking a stupid question here). Is there an easy way to do this from command line?
> 
> Edit: found this on *stackoverflow. `git push origin --force localBranchName:remoteBranchName` . Let me know if this is what I need to do.
> 
> Btw, yesterday I got myself an Xbox! I'll mod it as soon as I can; you can hit me up on Discord if you need sth to be tested on OG HW.

`git push -f` should be sufficient since the upstream branch is probably already set. If it's not sufficient, the error message should tell you exactly what to do.

There are also at least a handful of programmers on the Discord channel, so you can always ask there in the future and likely get help.

abaire commented:
> I might have messed up things a little, but I finally had to use `git push origin --force main:expand-blend-tests` and I think it's done now 👍 Thanks for your help and patience xD

Thanks! Now the problem is that the commit message isn't clear, so you'll want to do `git commit --amend` to edit at least the first line of the message (I think your original one was `Added tests for unsigned blending equations`) and then `git push -f` again to force it to update. 

In this case, I don't think any of the messages from the intermediate commits are particularly useful, so you can probably just wipe out the entire message while you're editing it and just use your original commit's message.

polymetal0 commented:
> Thanks! Now the problem is that the commit message isn't clear, so you'll want to do `git commit --amend` to edit at least the first line of the message (I think your original one was `Added tests for unsigned blending equations`) and then `git push -f` again to force it to update.
> 
> In this case, I don't think any of the messages from the intermediate commits are particularly useful, so you can probably just wipe out the entire message while you're editing it and just use your original commit's message.

I did what you suggested, let me know if there's anything else left.

abaire commented:
> > Thanks! Now the problem is that the commit message isn't clear, so you'll want to do `git commit --amend` to edit at least the first line of the message (I think your original one was `Added tests for unsigned blending equations`) and then `git push -f` again to force it to update.
> > In this case, I don't think any of the messages from the intermediate commits are particularly useful, so you can probably just wipe out the entire message while you're editing it and just use your original commit's message.
> 
> I did what you suggested, let me know if there's anything else left.

Looks perfect, thanks for the contribution!

I patched the change when I regenerated the HW goldens yesterday, you can find the results here: https://github.com/abaire/nxdk_pgraph_tests_golden_results/wiki/Results-Blend_tests

I will probably do a followup to render some boxes with solid alpha so we can also evaluate the effects on RGB for the images where nothing interesting is rendered (unless you'd like to do that as a second PR).

polymetal0 commented:
> Looks perfect, thanks for the contribution!
> 
> I patched the change when I regenerated the HW goldens yesterday, you can find the results here: https://github.com/abaire/nxdk_pgraph_tests_golden_results/wiki/Results-Blend_tests
> 
> I will probably do a followup to render some boxes with solid alpha so we can also evaluate the effects on RGB for the images where nothing interesting is rendered (unless you'd like to do that as a second PR).

For now, I'll get back to the homework you left me back in the PR I uploaded in the xemu repo 👍
